source/simulator/helper_functions.cc

```diff
@@ -2403,6 +2404,235 @@ namespace aspect
+                {
+                  phi_rho[k] = fe_values[introspection.extractors.compositional_fields[0]].value(k,q);
+                  phi_eta[k] = fe_values[introspection.extractors.compositional_fields[1]].value(k,q);
+                }
+
+              for (unsigned int i = 0; i<dofs_per_cell; ++i)
+                {
+                  // Calculate the density kernel
+                  local_rhs(i) += -(gravity * velocity_adjoint) * phi_rho[i] * fe_values.JxW(q);
+
+                  // Calculate the viscosity kernel
+                  local_rhs(i) += (2 * eta * strain_rate_forward * strain_rate_adjoint) * phi_eta[i] * fe_values.JxW(q);
+                  // it's fine to add both of them because only one of them at a time is nonzero
+
+                  for (unsigned int j = 0; j<dofs_per_cell; j++)
+                    {
+                      // Assemble Mass Matrix for the first compositional field. the mass
+                      // matrix for the second compositional field would look exactly the
+                      // same, so we can avoid building it for efficiency. (in fact we
+                      // also would have to deal with the fact that no sparsity pattern
+                      // and memory is allocated for the second compositional field; but
+                      // because we don't build the matrix, we don't care here.)
+                      local_mass_matrix(i,j) += fe_values[introspection.extractors.compositional_fields[0]].value(i,q) *
+                                                fe_values[introspection.extractors.compositional_fields[0]].value(j,q) *
+                                                fe_values.JxW(q);
+                    }
+                }
+            }
+
+
+          // SURFACE TERM
+          // TODO: the solution at the surface does not agree with the benchmark.
+          // this could be because of the calculation here or the assembly of the RHS in assemblers/adjoint.cc
+
+          // see if the cell is at the *top* boundary and if so start surface term calculation
+          unsigned int top_face_idx = numbers::invalid_unsigned_int;
+          for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+            if (cell->at_boundary(f) && geometry_model->depth (cell->face(f)->center()) < cell->face(f)->minimum_vertex_distance()/3)
+              {
+                top_face_idx = f;
+                break;
+              }
+
+          if (top_face_idx != numbers::invalid_unsigned_int)
+            {
+              fe_face_values.reinit (cell,top_face_idx);
+
+              // Evaluate the material properties and the solution at the cell boundary
+              MaterialModel::MaterialModelInputs<dim> in_face(fe_face_values, cell, introspection, solution);
+              MaterialModel::MaterialModelOutputs<dim> out_face(fe_face_values.n_quadrature_points, introspection.n_compositional_fields);
+              material_model->evaluate(in_face, out_face);
```

Comment:
average material model if necessary

Reply:
This currently doesn't work (see TODO in assembly). I've added a note about this in the code.